pkg/resource/aggregator.go

```diff
@@ -0,0 +1,215 @@
+
+			// read the container memory limits
+			for _, container := range nodeSet.PodTemplate.Spec.Containers {
+				if container.Name == esv1.ElasticsearchContainerName {
+					mem = *container.Resources.Limits.Memory()
+
+					// if not, fallback to twice the max JVM heap size
+					if mem.IsZero() {
+						for _, envVar := range container.Env {
+							if envVar.Name == essettings.EnvEsJavaOpts {
+								mem, err = memFromJavaOpts(envVar.Value)
+								if err != nil {
+									return resource.Quantity{}, err
+								}
+							}
+						}
+					}
+				}
+			}
+
+			// if not, fallback to the default limits
+			if mem.IsZero() {
+				mem = nodespec.DefaultMemoryLimits
+			}
+
+			total.Add(multiply(mem, nodeSet.Count))
+			log.V(1).Info("Collecting", "es_name", es.Name, "memory", mem.String(), "count", nodeSet.Count)
+		}
+	}
+
+	return total, nil
+}
+
+func (a Aggregator) aggregateKibanaMemory() (resource.Quantity, error) {
+	var kbList kbv1.KibanaList
+	err := a.client.List(&kbList)
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+
+	var total resource.Quantity
+	for _, kb := range kbList.Items {
+		var mem resource.Quantity
+
+		// read the container memory limits
+		for _, container := range kb.Spec.PodTemplate.Spec.Containers {
+			if container.Name == kbv1.KibanaContainerName {
+				mem = *container.Resources.Limits.Memory()
+
+				// if not, fallback to the max JVM heap size
+				if mem.IsZero() {
+					for _, envVar := range container.Env {
+						if envVar.Name == kbconfig.EnvNodeOpts {
+							mem, err = memFromNodeOptions(envVar.Value)
+							if err != nil {
+								return resource.Quantity{}, err
+							}
+						}
+					}
+				}
+			}
+		}
+
+		// if not, fallback to the default limits
+		if mem.IsZero() {
+			mem = kbpod.DefaultMemoryLimits
+		}
+
+		total.Add(multiply(mem, kb.Spec.Count))
+		log.V(1).Info("Collecting", "kibana_name", kb.Name, "memory", mem.String(), "count", kb.Spec.Count)
+	}
+
+	return total, nil
+}
+
+func (a Aggregator) aggregateApmServerMemory() (resource.Quantity, error) {
+	var asList asv1.ApmServerList
+	err := a.client.List(&asList)
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+
+	var total resource.Quantity
+	for _, as := range asList.Items {
+		var mem resource.Quantity
+
+		// read the container memory limits
+		for _, container := range as.Spec.PodTemplate.Spec.Containers {
+			if container.Name == asv1.ApmServerContainerName {
+				mem = *container.Resources.Limits.Memory()
+			}
+		}
+
+		// if not, fallback to the default limits
+		if mem.IsZero() {
+			mem = apmserver.DefaultMemoryLimits
+		}
+
+		total.Add(multiply(mem, as.Spec.Count))
+		log.V(1).Info("Collecting", "as_name", as.Name, "memory", mem.String(), "count", as.Spec.Count)
+	}
+
+	return total, nil
+}
+
+// maxHeapSizePattern is the pattern to extract the max Java heap size (-Xmx<size>[g|G|m|M|k|K])
+const maxHeapSizePattern = "-Xmx([0-9]*)([gGmMkK]*)"
+
+var maxHeapSizeRe = regexp.MustCompile(maxHeapSizePattern)
```

Comment:
nit: I think it's fine to have a single variable here:
```
var maxHeapSizeRe = regexp.MustCompile("var maxHeapSizeRe = regexp.MustCompile")
```
but totally fine if you prefer having 2 👍